Migrate pygstc to asyncio based sockets.

jameshilliard opened the pull request:
This should allow pygstc to be used properly in a non-blocking manner in addition to providing a cleaner abstraction over the TCP socket operations.

I also fixed the python tests to run properly in github actions.

michaelgruner commented:
I love this. I'll review calmly next week. Thanks for your multiple submissions.

emmadrigal commented on libgstc/python/setup.py:
```diff
@@ -44,10 +44,10 @@
-                 'Programming Language :: Python :: 3.6',
-                 'Programming Language :: Python :: 3.7'],
-    python_requires='>=3.5',
+                 'Programming Language :: Python :: 3.7',
+                 'Programming Language :: Python :: 3.8',
+                 'Programming Language :: Python :: 3.9'],
+    python_requires='>=3.7',
```
Does `asyncio` not support `3.5`, otherwise we might be breaking support

jameshilliard replied:
Well basic asyncio does support `3.5`, however [specific features](https://github.com/RidgeRun/gstd-1.x/pull/252#discussion_r721497115) I used have a minimum version of `3.7`.

jameshilliard replied:
FYI python `3.5` is already end of life, and `3.6` will be end of life [within a couple of months](https://www.python.org/dev/peps/pep-0494/#lifespan) looks like.

jameshilliard replied:
python 3.6 is [now end of life](https://www.python.org/dev/peps/pep-0494/#lifespan) so requiring python 3.7 shouldn't be an issue anymore right?

emmadrigal commented on tests/libgstc/Makefile.am:
```diff
@@ -1,5 +1,5 @@
 SUBDIRS=c
 
 # FIXME Disable pygstc checks
-#SUBDIRS = python
+SUBDIRS = python
```
Is the `FIXME` still valid?

jameshilliard replied:
Shouldn't be needed anymore, removed.

emmadrigal commented on tests/libgstc/python/test_libgstc_python_signal_disconnect.py:
```diff
@@ -28,41 +28,37 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 # OF THE POSSIBILITY OF SUCH DAMAGE.
-
+import asyncio
```
Leave a space after the License

jameshilliard replied:
Space added.

jameshilliard commented on libgstc/python/pygstc/tcp.py:
```diff
@@ -28,9 +28,9 @@
-import select
+import asyncio
 import socket
+from contextlib import asynccontextmanager
```
This feature has a minimum python version requirement of `3.7`. I can probably rework this to not require it but it makes socket cleanup a bit cleaner IMO.

jameshilliard commented on .github/workflows/main.yml:
```diff
@@ -11,7 +11,7 @@ on:
 jobs:
   build:
-    runs-on: ubuntu-18.04
+    runs-on: ubuntu-20.04
```
That was due to 18.04 shipping with python `3.6` while I used features here requiring a python `3.7` minimum.

jameshilliard commented:
I've rebased this however there's some failures unrelated to the changes in this pull request, I reworked the python test runner in #278 to fix the tests themselves(although some appear to still fail due to gstd bugs however).

jameshilliard commented:
I've rebased this on top of #278 which fixes most tests.

migueltaylor commented:
The main issue I see with this is that we currently support python >= 3.5. I could see migrating the client to python 3.6 but not to python 3.7 just yet. A lot of GstD users run it on NVIDIA JetPack and the latest version 4.6 still uses Ubuntu 18.04 and python 3.6. Unless there is an easy to maintain way to provide backwards compatibly I think we will have to wait a while to include this change.

jameshilliard commented:
> The main issue I see with this is that we currently support python >= 3.5.

Hmm, is it just NVIDIA JetPack using EOL python versions? 3.5 has been deprecated for a while now and 3.6 has been for a few months. This just updates the client library so I think if anyone need to use an older python they can just not install the new asyncio version probably right?

> A lot of GstD users run it on NVIDIA JetPack and the latest version 4.6 still uses Ubuntu 18.04 and python 3.6. Unless there is an easy to maintain way to provide backwards compatibly I think we will have to wait a while to include this change.

Can't they just do `apt install python3.7-dev`? Seems to work fine for me on a normal ubuntu 18.04 system at least.

TopperBG commented:
Keep in mind many Yocto projects, where change of packet, specialy Python is not as easy as apt install.

jameshilliard commented:
Rebased this just now, does this look good to merge now that python 3.6 has been end of life for a long time?
